pkg/auth/auth.go

```diff
@@ -28,18 +29,66 @@ import (
 	yamlUtils "github.com/kubeapps/kubeapps/pkg/yaml"
 )
 
-// UserAuth contains information to check user permissions
-type UserAuth struct {
-	authCli      authorizationv1.AuthorizationV1Interface
-	discoveryCli discovery.DiscoveryInterface
-}
-
 type resource struct {
 	APIVersion string
 	Kind       string
 	Namespace  string
 }
 
+type k8sAuthInterface interface {
+	Validate() error
+	GetResourceList(groupVersion string) (*metav1.APIResourceList, error)
+	CanI(verb, group, resource, namespace string) (bool, error)
+}
+
+type k8sAuth struct {
+	AuthCli      authorizationv1.AuthorizationV1Interface
+	DiscoveryCli discovery.DiscoveryInterface
+}
+
+func (u k8sAuth) Validate() error {
+	_, err := u.AuthCli.SelfSubjectRulesReviews().Create(&authorizationapi.SelfSubjectRulesReview{
+		Spec: authorizationapi.SelfSubjectRulesReviewSpec{
+			Namespace: "default",
+		},
+	})
+	return err
+}
+
+func (u k8sAuth) GetResourceList(groupVersion string) (*metav1.APIResourceList, error) {
+	return u.DiscoveryCli.ServerResourcesForGroupVersion(groupVersion)
+}
+
+func (u k8sAuth) CanI(verb, group, resource, namespace string) (bool, error) {
+	res, err := u.AuthCli.SelfSubjectAccessReviews().Create(&authorizationapi.SelfSubjectAccessReview{
+		Spec: authorizationapi.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationapi.ResourceAttributes{
+				Group:     group,
+				Resource:  resource,
+				Verb:      verb,
+				Namespace: namespace,
+			},
+		},
+	})
+	if err != nil {
+		return false, err
+	}
+	return res.Status.Allowed, nil
+}
+
+// UserAuth contains information to check user permissions
+type UserAuth struct {
+	k8sAuth k8sAuthInterface
```

Comment:
why not export k8sAuthInterface and use it directly?

Reply:
I think we are not understanding each other with the interface thing. There are two things we can mock here:

 - The whole UserAuth functionality. If we do that we can test whoever uses this pkg. In this case the user of this pkg is the `cmd/handler` that doesn't have tests (yet) so creating a `UserAuthInterface` and `UserAuthFake` right now won't be useful. (this is what I think you are suggesting, please correct me if I am wrong)
 - The K8s calls to the API. That way we don't need to rely too much in the current fake-auth and fake-discovery endpoints which are not complete enough to fully test the functionality of this package.

I did the second in order to be able to cover as much as possible the code of this pkg. The more code/functionality we include in the interface the less we can test.

Reply:
Okay, I see what you mean now, that makes sense. One thing you could do is use github.com/stretchr/testify/mock in the fake implementation so that you can change the response of `GetForbiddenActions` during the tests.

Reply:
I checked the package but the method mocked (`CanI`) is simple enough to have the same result than using `testObj.On("CanI", mock.Anything).Return(true, nil)`. I don't need to use that package for the moment to answer with a positive answer.